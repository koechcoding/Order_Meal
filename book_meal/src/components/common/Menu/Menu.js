import React from 'react';
import PropTypes from 'prop-types';
import MenuItem from './components/MenuItem';

