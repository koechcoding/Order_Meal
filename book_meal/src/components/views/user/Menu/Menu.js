import React from 'react';
import PropTypes from 'prop-types';
import axios from 'src/axios';
import Content from 'src/components/common/Content';
import MealMenu from 'src/components/common/Menu';
import OrderModal from './components/Order';

class Menu extends React.Component {

    state = {
        data: {},
        currentMenu: 0,
    }

    componentWillMount() {
        this.fetchMenus();
    }

    fetchMenus = () => {
        this.props.setLoading(true);
        axios.get('/menus?time=all').then(({ data }) => {
            this.setState({
                ...this.state,
                data
            });
            this.props.setLoading(false);
        }).catch(({ response }) =>  {
            this.props.setLoading(false);
        })
    }

    toggleOrder = () => {
        this.setState({
            ...this.state,
            orderIsOpen: !this.state.orderIsOpen
        });
    }

    onOrder = (item, menu) => {
        this.setState({
            ...this.state,
            toOrder: item,
            orderMenu: menu,
            orderIsOpen: !this.state.orderIsOpen
        });
    }

    onFilter = (e) => {
        this.setState({
            ...this.state,
            currentMenu: Number(e.target.value)
        });
    }

    onReset = () => {
        this.setState({
            ...this.state,
            currentMenu: 0
        })
    }

    render() {
        let { menus = [] } = this.state.data;
        const { toOrder, orderMenu, orderIsOpen, currentMenu } = this.state;

        let filtered = menus.filter((menu) => {
            if (currentMenu === 0) {
                return true;
            }
            return menu.id === currentMenu;
        });

        const contentTop = (
            <div className="col-12 mb-2 pr-0 pr-sm-2">
                <h5 className="d-inline-block">Today's Menu</h5>
            </div>
        );

        const contentFilter = (
            <div className="row mb-3">
                <div className="col-3 col-sm-2 col-lg-1 text-right">
                    <label className="pt-2">Menu:</label>
                </div>
                <div className="col-9 col-sm-7 col-lg-4 col-xl-3 pl-0 pr-1">
                    <select onChange={this.onFilter} className="form-control">
                        <option value="0">-- All --</option>
                        {menus.map((menu) => (
                            <option key={menu.id} value={menu.id}>{menu.name}</option>
                        ))}
                    </select>
                </div>
                {currentMenu !== 0 &&
                    <div className="animated fadeIn btn-reset col-12 col-sm-1 col-lg-1 pt-3 pt-sm-0 pl-0">
                        <button onClick={this.onReset} className="btn btn-secondary d-block m-auto">Reset </button>
                    </div>
                }
            </div>
        );

        return (
            <Content 
                {...this.props}
                contentTop={contentTop} 
                contentFilter={contentFilter}>
                {filtered.map((menu) => 
                    <MealMenu 
                        menu={menu}
                        key={menu.id} 
                        onOrder={this.onOrder}
                        items={menu.menu_items} 
                    />)
                }
                <OrderModal 
                    {...this.props}
                    size="sm"
                    item={toOrder} 
                    menu={orderMenu}
                    isOpen={orderIsOpen} 
                    toggle={this.toggleOrder}/>

            </Content>
        );
    }
}

Menu.propTypes = {
    setLoading: PropTypes.func.isRequired
}

export default Menu;