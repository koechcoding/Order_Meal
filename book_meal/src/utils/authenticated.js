export default function authenticated(){
    try{
        return JSON.parse(localStorage.getItem('user'));
    } catch(err){
        
    }
}