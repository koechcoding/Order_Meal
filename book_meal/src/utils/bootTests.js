import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
Enzyme.configure({ adapter: new Adapter()});

global.window = {}
import localStorage from 'mock-local-Storage';
window.localStorage = global.localStorage

export default Enzyme;
