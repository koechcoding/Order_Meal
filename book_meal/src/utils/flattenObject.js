export default function flattenObject(obj){
    let result = {}
    for (const key in obj){
        if((typeof obj[key])=== 'object' &&  obj[key]  !== null) {
            let flattened = flattenObject(obj[key])
            for(const key2 in flattened){
                result[key + "." + key2] = flattened[key2];
            }
        }else{
            result[key] = obj[key];
        }
    }
    return result;
}