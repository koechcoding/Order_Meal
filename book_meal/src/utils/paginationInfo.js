export default function paginationInfo(page){

    if(!page){
        return{}
    }

    let currentPage = 1
    if(page.next_page)
         currentPage = page.next_page -1;
    else if (page.prev_page)
        currentPage = page.prev_page + 1;

    let pageCount = Math.ceil(page.total / page.per_paage);
    if(Number.isNaN(pageCount) || pageCount === 0)
      pageCount = 1; 

    return {
        pageCount,
        currentPage,
        total: page.total,
        perPage: page.per_page,
        hasNext: page.has_next,
        hasPrev: page.has_prev,
        nextPage: page.next_page,
        prevPage: page.prev_page,
        currentCount: page.current_count,
    };
}