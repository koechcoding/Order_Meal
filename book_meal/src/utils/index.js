import singleError from './singleError';
import paginationInfo from './paginationInfo';
import authenticated from './authenticated';
import flattenObject from './flattenObject';

export { 
    singleError, 
    paginationInfo, 
    authenticated,
    flattenObject
}