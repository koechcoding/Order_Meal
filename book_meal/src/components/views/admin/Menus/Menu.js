import React from 'react';
import PropTypes from 'prop-types';
import { Alert, 
    ButtonDropdown,  
    DropdownToggle,
    DropdownMenu,
    DropdownItem
} from 'reactstrap';
import Filter from 'src/components/common/Filter';
import Content from 'src/components/common/Content';
import MenusTable from './components/MenusTable';
import CreateModal from './components/Create';
import EditModal from './components/Edit';
import DeleteModal from './components/Delete';
import MenuTypes from './components/MenuTypes';
import axios from 'src/axios';
import './styles.css';

import { singleError, paginationInfo } from 'src/utils';


class Menus extends React.Component {

    bubbleBlocked = false;

    constructor(props){
        super(props);
        this.state = {
            page: 1,
            data: {},
            search: '',
            perPage: 5,
            deleteIsOpen: false,
            editIsOpen: false,
            menuTypesIsOpen: false,
            toDelete: {},
            toEdit: {}
        }
    }
    componentWillMount() {
        this.fetchMenus()
    }

    fetchMenus = (config = {}) => {
        let { 
            page = this.state.page, 
            perPage = this.state.perPage,
            search = this.state.search, 
        } = config;
        const link = `/menu-items?page=${page}&search=${search}&per_page=${perPage}&time=today`;
        this.props.setLoading(true);

        axios.auth();
        axios.get(link, this.state).then(({ data }) => {
            const pageInfo = paginationInfo(data);
            this.setState({
                ...this.state,
                page: pageInfo.currentPage,
                data,
            });
            this.props.setLoading(false);

            if (pageInfo.currentCount === 0 && pageInfo.currentPage !== 1) {
                this.fetchMenus({
                    page: pageInfo.currentPage - 1
                })
            }
        }).catch(({ response }) => {
            this.setState({
                ...this.state,
                error: response,
            })
            this.props.setLoading(false);
        })
    }

    blockBubbling = () => {
        /** 
         * workaround e.stopPropagation()
         * @see https://github.com/facebook/react/issues/1691
         */
        this.bubbleBlocked = true;
        setTimeout(() => {
            this.bubbleBlocked = false;
        }, 500);

    }
    toggleMenuTypes = (e) => {
        this.blockBubbling();
        this.setState({
            ...this.state,
            manageIsOpen: false,
            menuTypesIsOpen: !this.state.menuTypesIsOpen
        });
    }
    toggleManage = (e) => {
        if (this.bubbleBlocked) 
            return;
        this.setState({
            ...this.state,
            manageIsOpen: !this.state.manageIsOpen,
        });
    }
    toggleManage = (e) => {
        if (this.bubbleBlocked) 
            return;
        this.setState({
            ...this.state,
            manageIsOpen: !this.state.manageIsOpen,
        });
    }
     
    toggleCreate = (e) => {
        this.blockBubbling();
        this.setState({
            ...this.state,
            manageIsOpen: false,
            createIsOpen: !this.state.createIsOpen,
        });
    }
    toggleEdit = (menuItem) => {
        this.setState({
            ...this.state,
            toEdit: menuItem || {},
            editIsOpen: !this.state.editIsOpen
        });
    }

    toggleDelete = (menuItem) => {
        this.setState({
            ...this.state,
            toDelete: menuItem || {},
            deleteIsOpen: !this.state.deleteIsOpen
        });
    }

    onFilter = (text) => {
        this.setState({
            ...this.state,
            search: text,
        });
        this.fetchMenus({search: text});
    }


}