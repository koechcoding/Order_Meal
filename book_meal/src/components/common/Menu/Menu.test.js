import React from 'react';
import { shallow }  from 'enzyme';
import 'src/utils/bootTests'; // order matters!
import Menu from './Menu';
import Adapter from 'enzyme-adapter-react-16';

const items = [{
        id: 0,
        name: 'name',
        cost: 50,
        img_url: '#'
    },
    {
        id: 1,
        name: 'name2',
        cost: 50,
        img_url: '#'
    }
]

describe('<Menu />', () => {
    it('renders correctly', () => {
        const wrapper = shallow(<Menu items={items} onOrder={() => {}} setLoading={() => {}}/>);
        expect(wrapper).toMatchSnapshot();
    });
});