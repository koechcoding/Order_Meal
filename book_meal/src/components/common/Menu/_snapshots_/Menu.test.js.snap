// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Menu
    items={
        Array [
            Object {
              "cost": 50,
              "id": 0,
              "img_url": "#",
              "name": "name",
            },
            Object {
              "cost": 50,
              "id": 1,
              "img_url": "#",
              "name": "name2",
            },
          ]
    }
    onOrder={[Function]}
    setLoading={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h5
          className="text-center mt-2"
/>,
        <div
          className="row"
>
          <MenuItem
                    item={
                              Object {
                                        "cost": 50,
                                        "id": 0,
                                        "img_url": "#",
                                        "name": "name",
                                      }
                    }
                    menu={Object {}}
                    onOrder={[Function]}
          />
          <MenuItem
                    item={
                              Object {
                                        "cost": 50,
                                        "id": 1,
                                        "img_url": "#",
                                        "name": "name2",
                                      }
                    }
                    menu={Object {}}
                    onOrder={[Function]}
          />
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "text-center mt-2",
        },
        "ref": null,
        "rendered": null,
        "type": "h5",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <MenuItem
              item={
                            Object {
                                          "cost": 50,
                                          "id": 0,
                                          "img_url": "#",
                                          "name": "name",
                                        }
              }
              menu={Object {}}
              onOrder={[Function]}
/>,
            <MenuItem
              item={
                            Object {
                                          "cost": 50,
                                          "id": 1,
                                          "img_url": "#",
                                          "name": "name2",
                                        }
              }
              menu={Object {}}
              onOrder={[Function]}
/>,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "function",
            "props": Object {
              "item": Object {
                "cost": 50,
                "id": 0,
                "img_url": "#",
                "name": "name",
              },
              "menu": Object {},
              "onOrder": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "function",
            "props": Object {
              "item": Object {
                "cost": 50,
                "id": 1,
                "img_url": "#",
                "name": "name2",
              },
              "menu": Object {},
              "onOrder": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h5
            className="text-center mt-2"
/>,
          <div
            className="row"
>
            <MenuItem
                        item={
                                    Object {
                                                "cost": 50,
                                                "id": 0,
                                                "img_url": "#",
                                                "name": "name",
                                              }
                        }
                        menu={Object {}}
                        onOrder={[Function]}
            />
            <MenuItem
                        item={
                                    Object {
                                                "cost": 50,
                                                "id": 1,
                                                "img_url": "#",
                                                "name": "name2",
                                              }
                        }
                        menu={Object {}}
                        onOrder={[Function]}
            />
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "text-center mt-2",
          },
          "ref": null,
          "rendered": null,
          "type": "h5",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <MenuItem
                item={
                                Object {
                                                "cost": 50,
                                                "id": 0,
                                                "img_url": "#",
                                                "name": "name",
                                              }
                }
                menu={Object {}}
                onOrder={[Function]}
/>,
              <MenuItem
                item={
                                Object {
                                                "cost": 50,
                                                "id": 1,
                                                "img_url": "#",
                                                "name": "name2",
                                              }
                }
                menu={Object {}}
                onOrder={[Function]}
/>,
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "function",
              "props": Object {
                "item": Object {
                  "cost": 50,
                  "id": 0,
                  "img_url": "#",
                  "name": "name",
                },
                "menu": Object {},
                "onOrder": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "function",
              "props": Object {
                "item": Object {
                  "cost": 50,
                  "id": 1,
                  "img_url": "#",
                  "name": "name2",
                },
                "menu": Object {},
                "onOrder": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;